ml-agents-envs/mlagents_envs/side_channel/side_channel_manager.py

```diff
@@ -0,0 +1,82 @@
+import uuid
+import struct
+from typing import Dict, Optional, List
+from mlagents_envs.side_channel import SideChannel, IncomingMessage
+from mlagents_envs.exception import UnityEnvironmentException
+from mlagents_envs.logging_util import get_logger
+
+
+class SideChannelManager:
+    def __init__(self, side_channels=Optional[List[SideChannel]]):
+        self._side_channels_dict = self._get_side_channels_dict(side_channels)
+
+    def process_side_channel_message(self, data: bytes) -> None:
+        """
+        Separates the data received from Python into individual messages for each
+        registered side channel and calls on_message_received on them.
+        :param data: The packed message sent by Unity
+        """
+        offset = 0
+        while offset < len(data):
+            try:
+                channel_id = uuid.UUID(bytes_le=bytes(data[offset : offset + 16]))
+                offset += 16
+                message_len, = struct.unpack_from("<i", data, offset)
+                offset = offset + 4
+                message_data = data[offset : offset + message_len]
+                offset = offset + message_len
+            except Exception:
+                raise UnityEnvironmentException(
+                    "There was a problem reading a message in a SideChannel. "
+                    "Please make sure the version of MLAgents in Unity is "
+                    "compatible with the Python version."
+                )
+            if len(message_data) != message_len:
+                raise UnityEnvironmentException(
+                    "The message received by the side channel {0} was "
+                    "unexpectedly short. Make sure your Unity Environment "
+                    "sending side channel data properly.".format(channel_id)
+                )
+            if channel_id in self._side_channels_dict:
+                incoming_message = IncomingMessage(message_data)
+                self._side_channels_dict[channel_id].on_message_received(
+                    incoming_message
+                )
+            else:
+                get_logger(__name__).warning(
+                    "Unknown side channel data received. Channel type "
+                    ": {0}.".format(channel_id)
+                )
+
+    def generate_side_channel_messages(self) -> bytearray:
+        """
+        Gathers the messages that the registered side channels will send to Unity
+        and combines them into a single message ready to be sent.
+        """
+        result = bytearray()
+        for channel_id, channel in self._side_channels_dict.items():
+            for message in channel.message_queue:
```

Comment:
Not really an issue for this PR, but it's surprising to me that `message_queue` isn't actually a queue.